src/gc.c

```diff
@@ -911,7 +1041,19 @@ static inline jl_taggedvalue_t *reset_page(const jl_gc_pool_t *p, jl_gc_pagemeta
     memset(pg->ages, 0, GC_PAGE_SZ / 8 / p->osize + 1);
     jl_taggedvalue_t *beg = (jl_taggedvalue_t*)(pg->data + GC_PAGE_OFFSET);
     jl_taggedvalue_t *next = (jl_taggedvalue_t*)pg->data;
-    next->next = fl;
+    if (fl == NULL) {
+        next->next = NULL;
+    }
+    else {
+        // Insert free page after first page.
+        // This prevents unnecessary fragmentation from multiple pages
+        // being allocated from at the same time. Instead, objects will
+        // only ever be allocated from the first object in the list.
```

Comment:
Oh, and this comment is still missing a mention of the relation to conservative scanning that I requested last time. That won't be clear for future reader out of the context of this PR.

Reply:
Noted. Will do that. I hadn't realized that you wanted me to make the connection to conservative scanning and thought you simply wanted the logic explained. But putting it also in the context of conservative scanning makes perfect sense.

Reply:
@rbehrends hmm, this comment has not yet Berlin addressed, has it?

Reply:
It has, the comment now explicitly notes that`jl_gc_internal_obj_base_ptr()` (which is part of the conservative marking machinery) relies on it.